ethpm/manifest.py

```diff
@@ -0,0 +1,149 @@
+                contract_type_data = self._create_contract_type(
+                    contract_name, contract, solc
+                )
+                setattr(
+                    self,
+                    "contract_types",
+                    {**self.contract_types, **contract_type_data},
+                )
+
+    @to_dict
+    def _create_contract_type(
+        self, contract_name: str, contract: str, solc: Dict[str, Any]
+    ) -> Generator[Tuple[str, Any], None, None]:
+        yield contract_name, self._generate_contract_type_dict(contract, solc)
+
+    @to_dict
+    def _generate_contract_type_dict(
+        self, contract: str, solc: Dict[str, Any]
+    ) -> Generator[Tuple[str, Any], None, None]:
+        yield "abi", json.loads(solc["contracts"][contract]["abi"])
+        yield "natspec", json.loads(solc["contracts"][contract]["devdoc"])
+
+    @to_dict
+    def _create_source(
+        self, contract_path: str
+    ) -> Generator[Tuple[str, Any], None, None]:
+        [ipfs_return_data] = self.ipfs_backend.pin_assets(
+            Path(self.solc_path).parent / contract_path
+        )
+        ipfs_hash = ipfs_return_data["Hash"]
+        ipfs_uri = create_ipfs_uri(ipfs_hash)
+        yield "./{0}".format(contract_path), ipfs_uri
+
+    def add_meta(self, **kwargs: Any) -> None:
+        """
+        Adds metadata to the `meta` key in this manifest.
+        Suggested kwarg types
+        - `authors`         List[str]
+        - `license`         str
+        - `description`     str
+        - `keywords`        List[str]
+        - `links`           Dict[str,str]
+            Recommended keys for `links`
+            - `website`
+            - `documentation`
+            - `repository`
+        """
+        self.meta = {key: value for key, value in kwargs.items() if value}
+
+    def pretty(self) -> str:
+        """
+        Returns a string of a pretty printed version of this manifest.
+        """
+        attr_dict = self._generate_attr_dict()
+        return json.dumps(attr_dict, indent=4, sort_keys=True)
+
+    def minified(self) -> str:
+        """
+        Returns a string of a minified version of this manifest.
+        """
+        attr_dict = self._generate_attr_dict()
+        return json.dumps(attr_dict, separators=(",", ":"), sort_keys=True)
+
+    # TODO implement `write_pretty_to_disk` w/o indenting ABI
+    def write_minified_to_disk(self) -> Path:
+        """
+        Returns a path leading to a minified version of this manifest written to disk.
+        Manifest is written to the same dir in which the `contracts/` dir is located.
+        """
+        manifest_contents = self.minified()
+        path = self.solc_path.parent / str(self.version + ".json")
+        path.write_text(manifest_contents)
+        return path
+
+    @to_dict
+    def _generate_attr_dict(self) -> Generator[Tuple[str, Any], Any, None]:
+        for key, value in self.__dict__.items():
```

Comment:
I think this is going to blow up on you.  I'd recommend just doing this explicitely.  There are a fixed number of top level keys and it should be reasonably sane to just list them all.